Move attention kernels from meta_registrations to fake_impls

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120682

This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

Differential Revision: [D54328457](https://our.internmc.facebook.com/intern/diff/D54328457)

davidberard98 committed:
Move attention kernels from meta_registrations to fake_impls

This PR is just code movement to make the code review easier. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

[ghstack-poisoned]

davidberard98 committed:
Update on "Move attention kernels from meta_registrations to fake_impls"


This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

[ghstack-poisoned]

davidberard98 committed:
Update on "Move attention kernels from meta_registrations to fake_impls"


This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

[ghstack-poisoned]

drisspg commented on torch/_subclasses/fake_impls.py:
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
+    max_seqlen_batch_k = key.size(2)
+
+    query_t = query.transpose(1, 2)
+    attention = torch.empty_like(query_t).transpose(1, 2)
```
dumb q, more just pattern matching why doesn't this need `convert_tensor`?

davidberard98 replied:
I think empty_like returns a fake tensor already, so it doesn't need to be wrapped (not really sure how this works, I just noticed it throws errors if I try to wrap it)

drisspg replied:
ohh duh  cause query_t is fake

drisspg commented on torch/_subclasses/fake_impls.py:
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
+    return_debug_mask = kwargs["return_debug_mask"]
+    # unused: value, dropout_p, is_causal, scale
+
+    def wrap_tensor(t, device):
```
nit, maybe use same name

drisspg commented on torch/_subclasses/fake_impls.py:
```diff
@@ -595,6 +596,255 @@ def convert(t, mem_fmt):
+    # Cuda Path
+    # attention = wrap_tensor(torch.empty_like(query, device="meta"), query.device)
+    # attention = wrap_tensor(torch.empty(query.size(), layout=query.layout, dtype=query.dtype, device="meta"), query.device)
+    attention = torch.empty_like(query)
```
dido here for dumb q

davidberard98 committed:
Update on "Move attention kernels from meta_registrations to fake_impls"


This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

[ghstack-poisoned]

davidberard98 committed:
Update on "Move attention kernels from meta_registrations to fake_impls"


This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

[ghstack-poisoned]

davidberard98 committed:
Update on "Move attention kernels from meta_registrations to fake_impls"


This PR is mostly just code movement to make the code review easier - AFAIK it should not change any functionality. The final goal is to remove the xfails for some of the test_fake opinfos for these ops. The opinfos are failing because the outputs can have mixed devices - we need to move them to fake_impls first before we can support mixed device returns.

This PR:
* Move the `_meta_registrations.py` implementations to `fake_impls.py`
* Change the function signature from taking explicit named variables to taking `{args, kwargs}` and normalizing them
* Wrap all the returned tensors in FakeTensors

Tests: relying on opinfos. I also checked `test_fake_*` for these tests (by removing x-fails and patching things until they passed) to verify general correctness.

[ghstack-poisoned]

drisspg reviewed (approved):
Looks good!

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54328457).
